codegen/src/types/composite_def.rs

```diff
@@ -0,0 +1,320 @@
+                .any(|tp| Some(tp.original_name.to_string()) == field.type_name)
+            {
+                let ty = type_gen.resolve_type(field.type_id);
+                if matches!(
+                    ty.type_def(),
+                    TypeDef::Primitive(
+                        TypeDefPrimitive::U8
+                            | TypeDefPrimitive::U16
+                            | TypeDefPrimitive::U32
+                            | TypeDefPrimitive::U64
+                            | TypeDefPrimitive::U128
+                    )
+                ) {
+                    derives.push_codec_compact_as()
+                }
+            }
+        }
+
+        let name = format_ident!("{}", ident.to_camel_case());
+
+        Self {
+            name,
+            kind: CompositeDefKind::Struct {
+                derives,
+                type_params,
+                field_visibility,
+            },
+            fields: fields_def,
+        }
+    }
+
+    /// Construct a definition which will generate code for an `enum` variant.
+    pub fn enum_variant_def(ident: &str, fields: CompositeDefFields) -> Self {
+        let name = format_ident!("{}", ident);
+        Self {
+            name,
+            kind: CompositeDefKind::EnumVariant,
+            fields,
+        }
+    }
+}
+
+impl quote::ToTokens for CompositeDef {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let name = &self.name;
+
+        let decl = match &self.kind {
+            CompositeDefKind::Struct {
+                derives,
+                type_params,
+                field_visibility,
+            } => {
+                let unused_phantom_marker = type_params.unused_params_phantom_data();
+                let fields = self.fields.field_tokens(
+                    field_visibility.as_ref(),
+                    unused_phantom_marker,
+                    true,
+                );
+
+                quote! {
+                    #derives
+                    pub struct #name #type_params #fields
+                }
+            }
+            CompositeDefKind::EnumVariant => {
+                let fields = self.fields.field_tokens(None, None, false);
+
+                quote! {
+                    #name #fields
+                }
+            }
+        };
+        tokens.extend(decl)
+    }
+}
+
+/// Which kind of composite type are we generating, either a standalone `struct` or an `enum`
+/// variant.
+#[derive(Debug)]
+pub enum CompositeDefKind {
+    /// Composite type comprising a Rust `struct`.
+    Struct {
+        derives: GeneratedTypeDerives,
+        type_params: TypeDefParameters,
+        field_visibility: Option<syn::Visibility>,
+    },
+    /// Comprises a variant of a Rust `enum`.
+    EnumVariant,
+}
+
+/// Encapsulates the composite fields, keeping the invariant that all fields are either named or
+/// unnamed.
+#[derive(Debug)]
+pub struct CompositeDefFields {
+    named: bool,
+    fields: Vec<CompositeDefField>,
+}
```

Comment:
I wonder whether a slightly better fitting type here would be something like:

```
enum CompositeDefFields {
   Named(Vec<(syn::Ident, CompositeDefField)>),
   Unnamed(Vec<CompositeDefField>)
}
```

And then remove the `name` from `CompositeDefField`. You wouldn't have to check a boolean or handle the option<names> in that case :)

Reply:
Indeed, that is precisely the type as it currently is: https://github.com/paritytech/subxt/blob/415e1222d23ca5b00a68128f63c74e4554780a4e/codegen/src/struct_def.rs#L40

It's been a while since I did this but I remember getting into difficulties with this type across it's different usage contexts, hence the relaxing of the strict type.

I will look again at this to see whether it can be improved.

Reply:
This change has now been applied, let me know if it looks good